File: src/main/java/com/qtech/bubbles/core/controllers/MouseController.java
package com.qtech.bubbles.core.controllers;

import com.qtech.bubbles.QBubbles;
import com.qtech.bubbles.event.MouseMotionEvent;
import com.qtech.bubbles.event.type.MouseEventType;

import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.event.MouseWheelEvent;

/**
 * @see KeyboardController
 * @see java.awt.event.MouseAdapter
 */
@SuppressWarnings("ConstantConditions")
public class MouseController extends MouseAdapter {
    private static final MouseController INSTANCE = new MouseController();
    private Point currentLocationOnScreen;
    private Point currentPoint;

    private MouseController() {

    }

    public static MouseController instance() {
        return INSTANCE;
    }

    @Override
    public void mouseClicked(MouseEvent e) {
        currentLocationOnScreen = e.getLocationOnScreen() != null ? e.getLocationOnScreen() : currentLocationOnScreen;
        currentPoint = e.getPoint() != null ? e.getPoint() : currentPoint;

        QBubbles.getEventBus().post(new com.qtech.bubbles.event.MouseEvent(QBubbles.getInstance(), this, e, MouseEventType.CLICK));
    }

    @Override
    public void mousePressed(MouseEvent e) {
        currentLocationOnScreen = e.getLocationOnScreen() != null ? e.getLocationOnScreen() : currentLocationOnScreen;
        currentPoint = e.getPoint() != null ? e.getPoint() : currentPoint;

        QBubbles.getEventBus().post(new com.qtech.bubbles.event.MouseEvent(QBubbles.getInstance(), this, e, MouseEventType.PRESS));
    }

    @Override
    public void mouseReleased(MouseEvent e) {
        currentLocationOnScreen = e.getLocationOnScreen() != null ? e.getLocationOnScreen() : currentLocationOnScreen;
        currentPoint = e.getPoint() != null ? e.getPoint() : currentPoint;

        QBubbles.getEventBus().post(new com.qtech.bubbles.event.MouseEvent(QBubbles.getInstance(), this, e, MouseEventType.RELEASE));
    }

    @Override
    public void mouseEntered(MouseEvent e) {
        currentLocationOnScreen = e.getLocationOnScreen() != null ? e.getLocationOnScreen() : currentLocationOnScreen;
        currentPoint = e.getPoint() != null ? e.getPoint() : currentPoint;

        QBubbles.getEventBus().post(new MouseMotionEvent(QBubbles.getInstance(), this, e, MouseEventType.ENTER));
    }

    @Override
    public void mouseExited(MouseEvent e) {
        currentLocationOnScreen = e.getLocationOnScreen() != null ? e.getLocationOnScreen() : currentLocationOnScreen;
        currentPoint = e.getPoint() != null ? e.getPoint() : currentPoint;

        QBubbles.getEventBus().post(new MouseMotionEvent(QBubbles.getInstance(), this, e, MouseEventType.LEAVE));
    }

    @Override
    public void mouseDragged(MouseEvent e) {
        currentLocationOnScreen = e.getLocationOnScreen() != null ? e.getLocationOnScreen() : currentLocationOnScreen;
        currentPoint = e.getPoint() != null ? e.getPoint() : currentPoint;

        QBubbles.getEventBus().post(new MouseMotionEvent(QBubbles.getInstance(), this, e, MouseEventType.DRAG));
    }

    @Override
    public void mouseMoved(MouseEvent e) {
        currentLocationOnScreen = e.getLocationOnScreen() != null ? e.getLocationOnScreen() : currentLocationOnScreen;
        currentPoint = e.getPoint() != null ? e.getPoint() : currentPoint;

        QBubbles.getEventBus().post(new MouseMotionEvent(QBubbles.getInstance(), this, e, MouseEventType.MOTION));
    }

    @Override
    public void mouseWheelMoved(MouseWheelEvent e) {
        currentLocationOnScreen = e.getLocationOnScreen() != null ? e.getLocationOnScreen() : currentLocationOnScreen;
        currentPoint = e.getPoint() != null ? e.getPoint() : currentPoint;
        QBubbles.getEventBus().post(new com.qtech.bubbles.event.MouseWheelEvent(QBubbles.getInstance(), this, e, MouseEventType.MOTION));

    }

    public Point getCurrentLocationOnScreen() {
        return currentLocationOnScreen;
    }

    public Point getCurrentPoint() {
        return currentPoint;
    }
}

File: src/main/kotlin/com/qtech/bubbles/core/controllers/MouseController.kt
package com.qtech.bubbles.core.controllers

import com.qtech.bubbles.BubbleBlaster
import com.qtech.bubbles.event.MouseMotionEvent
import com.qtech.bubbles.event.type.MouseEventType
import java.awt.Point
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.awt.event.MouseWheelEvent

/**
 * @see KeyboardController
 *
 * @see java.awt.event.MouseAdapter
 */
class MouseController private constructor() : MouseAdapter() {
    var currentLocationOnScreen: Point? = null
        private set
    var currentPoint: Point? = null
        private set

    override fun mouseClicked(e: MouseEvent) {
        currentLocationOnScreen = if (e.locationOnScreen != null) e.locationOnScreen else currentLocationOnScreen
        currentPoint = e.point
        BubbleBlaster.eventBus.post(com.qtech.bubbles.event.MouseEvent(BubbleBlaster.instance, this, e, MouseEventType.CLICK))
    }

    override fun mousePressed(e: MouseEvent) {
        currentLocationOnScreen = if (e.locationOnScreen != null) e.locationOnScreen else currentLocationOnScreen
        currentPoint = e.point
        BubbleBlaster.eventBus.post(com.qtech.bubbles.event.MouseEvent(BubbleBlaster.instance, this, e, MouseEventType.PRESS))
    }

    override fun mouseReleased(e: MouseEvent) {
        currentLocationOnScreen = if (e.locationOnScreen != null) e.locationOnScreen else currentLocationOnScreen
        currentPoint = e.point
        BubbleBlaster.eventBus.post(com.qtech.bubbles.event.MouseEvent(BubbleBlaster.instance, this, e, MouseEventType.RELEASE))
    }

    override fun mouseEntered(e: MouseEvent) {
        currentLocationOnScreen = if (e.locationOnScreen != null) e.locationOnScreen else currentLocationOnScreen
        currentPoint = e.point
        BubbleBlaster.eventBus.post(MouseMotionEvent(BubbleBlaster.instance, this, e, MouseEventType.ENTER))
    }

    override fun mouseExited(e: MouseEvent) {
        currentLocationOnScreen = if (e.locationOnScreen != null) e.locationOnScreen else currentLocationOnScreen
        currentPoint = e.point
        BubbleBlaster.eventBus.post(MouseMotionEvent(BubbleBlaster.instance, this, e, MouseEventType.LEAVE))
    }

    override fun mouseDragged(e: MouseEvent) {
        currentLocationOnScreen = if (e.locationOnScreen != null) e.locationOnScreen else currentLocationOnScreen
        currentPoint = e.point
        BubbleBlaster.eventBus.post(MouseMotionEvent(BubbleBlaster.instance, this, e, MouseEventType.DRAG))
    }

    override fun mouseMoved(e: MouseEvent) {
        currentLocationOnScreen = if (e.locationOnScreen != null) e.locationOnScreen else currentLocationOnScreen
        currentPoint = e.point
        BubbleBlaster.eventBus.post(MouseMotionEvent(BubbleBlaster.instance, this, e, MouseEventType.MOTION))
    }

    override fun mouseWheelMoved(e: MouseWheelEvent) {
        currentLocationOnScreen = if (e.locationOnScreen != null) e.locationOnScreen else currentLocationOnScreen
        currentPoint = e.point
        BubbleBlaster.eventBus.post(com.qtech.bubbles.event.MouseWheelEvent(BubbleBlaster.instance, this, e, MouseEventType.MOTION))
    }

    companion object {
        private val INSTANCE = MouseController()
        @JvmStatic
        fun instance(): MouseController {
            return INSTANCE
        }
    }
}